android/src/main/java/com/reactlibrary/RNSpotifyRemoteAppModule.java

```diff
@@ -69,48 +67,43 @@ public void onFailure(Throwable throwable) {
                         promise.reject(throwable);   
                     }
                 }
-                sendEvent("remoteDisconnected", null);
+                sendEvent(reactContext, "remoteDisconnected", null);
             }
         };
     }
 
-    @Override
-    @ReactMethod
-    public void __registerAsJSEventEmitter(int moduleId) {
-        RNEventEmitter.registerEventEmitterModule(this.reactContext, moduleId, this);
-    }
-
-    @Override
-    public void onNativeEvent(String eventName, Object... args) {
-        // Called when an event for this module is emitted from native code
-    }
-
-    @Override
-    public void onJSEvent(String eventName, Object... args) {
-        // Called when an event for this module is emitted from javascript
+    private void sendEvent(ReactApplicationContext reactContext,
+                      String eventName,
+                      Object params) {
+        reactContext
+            .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
+            .emit(eventName, params);
     }
 
-    @Override
-    public void onEvent(String eventName, Object... args) {
-        // Called when any event for this module is emitted
+    @ReactMethod
+    public void addListener(String eventName) {
```

Comment:
Since we know what event is being subscribed to here, we could more efficiently subscribe to events (i.e. in `handleOnConnect` we just subscribe to all interesting events.

Reply:
I will try to do that. However, we cannot detect if all event listeners is removed for a specific event since removeListeners only accept a `count` param: https://github.com/facebook/react-native/blob/4f1f42b0761bff7a68c85cae6de0c4c58c5031d5/Libraries/EventEmitter/NativeEventEmitter.js#L89

Reply:
Added lazy subscription (but not unsubscription yet)

Reply:
So the public native API does not have a method for unsubscription. So we may need to resort to adding the logic to JS side and add native method to be called from unsubscription. However, I wonder if it worth tracking those event since it might not be that bad with only several event listeners.

Reply:
yea, good point.  If it seems like a lot of work then it might not be worth it, I just wanted to avoid leaking event handlers if we could avoid it.

Reply:
@cjam Added lazy subscription/unsubscription for both Android and iOS :rocket: e55d6f258579426d2e4c081d860515adf6a62d18. I have tested a bit on Android and it seem to work. But as I don't know how to code ObjectiveC, if you can take a look it would be great!